h/static/scripts/config/identity.js

```diff
@@ -0,0 +1,139 @@
+var JWT_BEARER = 'urn:ietf:params:oauth:grant-type:jwt-bearer';
+
+var _authPromise = null;
+var _tokenPromise = null;
+var _tokenResponse = null;
+
+
+fetchToken.$inject = ['$http', 'jwtHelper', 'serviceUrl', 'session']
+function fetchToken  ( $http,   jwtHelper,   serviceUrl,   session ) {
+  var tokenUrl = new URL('token', serviceUrl).href;
+  var token = (_tokenResponse === null) ? null : _tokenResponse;
+
+  if (token === null || jwtHelper.isTokenExpired(token)) {
+    if (_tokenPromise === null) {
+      // Set up the token request data.
+      var data = {
+        assertion: session.state.id_token,
+        grant_type: JWT_BEARER
+      };
+
+      // Skip JWT authorization for the token request itself.
+      var config = {
+        skipAuthorization: true
+      };
+
+      // Make the request.
+      var request = $http.post(tokenUrl, data, config);
+
+      // Extract and save the response data.
+      _tokenPromise = request.then(function (response) {
+        _tokenPromise = null;
+        _tokenResponse = response.data;
+        return _tokenResponse;
+      });
+    }
+
+    // Return a promise of the access token.
+    return _tokenPromise;
+  } else {
+    // The token is available and not expired.
+    return token;
+  }
+}
+
+
+tokenGetter.$inject = ['$injector', 'config', 'serviceUrl'];
+function tokenGetter  ( $injector,   config,   serviceUrl ) {
+  var requestUrl = config.url;
+
+  // Only send the token on requests to the annotation storage service
+  // and only if it is not the token request itself.
+  if (requestUrl !== serviceUrl) {
+    if (requestUrl.slice(0, serviceUrl.length) === serviceUrl) {
+      return _authPromise
+        .then(function () {
+          return $injector.invoke(fetchToken);
+        })
+        .catch(function () {
+          return null;
+        });
+    }
+  }
+
+  return null;
+}
+
+
+configureIdentity.$inject = ['identityProvider', 'jwtInterceptorProvider'];
+function configureIdentity  ( identityProvider,   jwtInterceptorProvider ) {
+
+  identityProvider.checkAuthentication = [
+    '$injector', '$q', 'session', function($injector, $q, session) {
+      if (_authPromise === null) {
+        var deferred = $q.defer();
+        _authPromise = deferred.promise;
+
+        session.load().$promise
+          .then(function (data) {
+            if (data.userid) {
+              $q.when($injector.invoke(fetchToken))
```

Comment:
This could be simplified by having `fetchToken` return a promise even when it has a token already, i.e. `return Promise.resolve(token)`.